presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java

```diff
@@ -91,17 +90,14 @@ public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int
             }
 
             Analyzer analyzer = new Analyzer(parser, specialRules, specialTokens, ignoredRules);
-            Multimap<Integer, String> candidates = analyzer.process(currentState, currentToken.getTokenIndex(), context);
+            Result result = analyzer.process(currentState, currentToken.getTokenIndex(), context);
 
             // pick the candidate tokens associated largest token index processed (i.e., the path that consumed the most input)
-            String expected = candidates.asMap().entrySet().stream()
-                    .max(Comparator.comparing(Map.Entry::getKey))
-                    .get()
-                    .getValue().stream()
+            String expected = result.getExpected().stream()
                     .sorted()
                     .collect(Collectors.joining(", "));
 
-            message = format("mismatched input '%s'. Expecting: %s", ((Token) offendingSymbol).getText(), expected);
+            message = format("mismatched input '%s'. Expecting: %s", parser.getTokenStream().get(result.getErrorToken()).getText(), expected);
```

Comment:
Using `result.getErrorToken()` (`furthestToken`) there is a theoretical possibility we will say "mismatched" at quoting something else than ANTLR choked at. 

I replaced `parser.getTokenStream().get(result.getErrorToken())` with `((Token) offendingSymbol)` and it didnt have any noticable impact.
Maybe at least something like
```
if (((Token) offendingSymbol).getTokenIndex() != result.getErrorToken()) {
   log.warn ..
}
```
?

Reply:
It doesn't really matter. All we care about is that there is *some* syntax error and we want to produce a reasonable error. There's no false positive issue here where we might report an error where there is none.